Dockerfile

```diff
@@ -1,93 +1,131 @@
 ###########################################################
-# Base container with all necessary deps
-# Buster slim python 3.7 base image.
-FROM python:3.7-slim-buster as base
-ENV HTTP_PORT 8080
-
-RUN groupadd -r geoadmin \
-    && useradd -r -s /bin/false -g geoadmin geoadmin \
-    && mkdir -p /logs && chown geoadmin:geoadmin /logs \
-    # install relevent packages
-    && apt-get update \
-    && apt-get install -y binutils libproj-dev gdal-bin \
-    # the following line contains debug tools that can be removed later
+# Container that contains basic configurations used by all other containers
+# It should only contain variables that don't change or change very infrequently
+# so that the cache is not needlessly invalidated
+FROM python:3.9-slim-bullseye as base
+ENV HTTP_PORT=8080
+ENV USER=geoadmin
+ENV GROUP=geoadmin
+ENV INSTALL_DIR=/opt/service-stac
+
+RUN groupadd -r ${GROUP} \
+    && useradd -r -s /bin/false -g ${GROUP} ${USER} \
+    && mkdir -p ${INSTALL_DIR}/logs && chown ${USER}:${GROUP} ${INSTALL_DIR}/logs
+
+WORKDIR ${INSTALL_DIR}
+EXPOSE $HTTP_PORT
+# entrypoint is the manage command
+ENTRYPOINT ["python"]
+
+###########################################################
+# Container to perform tests/management/dev tasks
+FROM base as debug
+LABEL target=debug
+RUN apt-get update -qq \
+    && DEBIAN_FRONTEND=noninteractive apt-get install -y gdal-bin \
+    # dev dependencies
+    binutils libproj-dev \
+    # debug tools
     curl net-tools iputils-ping postgresql-client-common jq openssh-client \
     && apt-get clean \
     && rm -rf /var/lib/apt/lists/* \
     && pip3 install pipenv \
     && pipenv --version
 
-COPY Pipfile* /tmp/
-RUN cd /tmp && \
-    pipenv install --system --deploy --ignore-pipfile
+ENV PIPENV_VENV_IN_PROJECT=1
+COPY Pipfile.lock Pipfile ./
+RUN pipenv sync --dev
+
+# Is it safe for the executable files to be writable by the user who executes them? (At a first
+# glance, it seems to also work without this chown, but as I am not sure why this was added I left
+# it like this for the moment)
+COPY --chown=${USER}:${GROUP} .env.default ./
+COPY --chown=${USER}:${GROUP} spec/ spec/
+COPY --chown=${USER}:${GROUP} app/ app/
+
+# Activate virtualenv
+ENV VIRTUAL_ENV=${INSTALL_DIR}/.venv
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+# this is only used with the docker-compose setup within CI
+# to ensure that the app is only started once the DB container
+# is ready
+COPY ./wait-for-it.sh /app/
```

Comment:
I would move this after the installation of pip dependencies in line 37. This script is unlikely to change in opposition to the files in `spec` or `app` folders.